File: 03_use_cases/13_macos_cua/python/metacua/mouse.py
"""Synthetic mouse events via CGEvent."""

import enum
import math
import time
from typing import Optional, Tuple

from CoreFoundation import CFPreferencesCopyAppValue, kCFPreferencesAnyApplication
from Quartz import (
    CGEventCreate,
    CGEventCreateMouseEvent,
    CGEventCreateScrollWheelEvent,
    CGEventGetLocation,
    CGEventPost,
    CGEventSetIntegerValueField,
    CGEventSourceCreate,
    CGPointMake,
    kCGEventLeftMouseDown,
    kCGEventLeftMouseDragged,
    kCGEventLeftMouseUp,
    kCGEventMouseMoved,
    kCGEventOtherMouseDown,
    kCGEventOtherMouseDragged,
    kCGEventOtherMouseUp,
    kCGEventRightMouseDown,
    kCGEventRightMouseDragged,
    kCGEventRightMouseUp,
    kCGEventSourceStateCombinedSessionState,
    kCGHIDEventTap,
    kCGMouseButtonCenter,
    kCGMouseButtonLeft,
    kCGMouseButtonRight,
    kCGMouseEventClickState,
    kCGScrollEventUnitLine,
)

from .errors import CLIError


def _sleep_us(microseconds: int) -> None:
    time.sleep(microseconds / 1_000_000.0)


class MouseButton(enum.Enum):
    LEFT = "left"
    RIGHT = "right"
    CENTER = "center"

    @staticmethod
    def parse(raw: str) -> "MouseButton":
        low = raw.lower()
        if low in ("left", "l", "primary"):
            return MouseButton.LEFT
        if low in ("right", "r", "secondary"):
            return MouseButton.RIGHT
        if low in ("center", "middle", "c", "m"):
            return MouseButton.CENTER
        raise CLIError(f"--mouse-button must be left|right|center (got '{raw}')", code=2)

    @property
    def cg_button(self) -> int:
        return {
            MouseButton.LEFT: kCGMouseButtonLeft,
            MouseButton.RIGHT: kCGMouseButtonRight,
            MouseButton.CENTER: kCGMouseButtonCenter,
        }[self]

    @property
    def click_types(self) -> Tuple[int, int]:
        """Down / up event types for this button."""
        return {
            MouseButton.LEFT: (kCGEventLeftMouseDown, kCGEventLeftMouseUp),
            MouseButton.RIGHT: (kCGEventRightMouseDown, kCGEventRightMouseUp),
            MouseButton.CENTER: (kCGEventOtherMouseDown, kCGEventOtherMouseUp),
        }[self]

    @property
    def drag_type(self) -> int:
        return {
            MouseButton.LEFT: kCGEventLeftMouseDragged,
            MouseButton.RIGHT: kCGEventRightMouseDragged,
            MouseButton.CENTER: kCGEventOtherMouseDragged,
        }[self]


class ScrollDirection(enum.Enum):
    UP = "up"
    DOWN = "down"
    LEFT = "left"
    RIGHT = "right"

    @staticmethod
    def parse(raw: str) -> "ScrollDirection":
        try:
            return ScrollDirection(raw.lower())
        except ValueError:
            raise CLIError(f"--direction must be up|down|left|right (got '{raw}')", code=2)


def event_source():
    """Shared event source for all generated events."""
    return CGEventSourceCreate(kCGEventSourceStateCombinedSessionState)


def _post(event) -> None:
    if event is not None:
        CGEventPost(kCGHIDEventTap, event)


def move_mouse(point, source) -> None:
    """Warp the cursor to a point in one event (instant; used per-step internally)."""
    _post(
        CGEventCreateMouseEvent(
            source, kCGEventMouseMoved, CGPointMake(point.x, point.y), kCGMouseButtonLeft
        )
    )


def current_mouse_location():
    """Current cursor location in global (top-left origin) coordinates."""
    event = CGEventCreate(None)
    if event is None:
        return CGPointMake(0, 0)
    return CGEventGetLocation(event)


def smooth_move(target, source) -> None:
    """Glide the cursor to `target` along an eased path instead of teleporting."""
    start = current_mouse_location()
    dx = target.x - start.x
    dy = target.y - start.y
    distance = math.sqrt(dx * dx + dy * dy)
    if distance < 1:
        move_mouse(target, source)
        return
    steps = max(12, min(60, int(distance / 12)))
    for i in range(1, steps + 1):
        t = i / steps
        # ease-in-out
        e = 2 * t * t if t < 0.5 else 1 - math.pow(-2 * t + 2, 2) / 2
        move_mouse(CGPointMake(start.x + dx * e, start.y + dy * e), source)
        _sleep_us(6_000)


def perform_click(point, button: MouseButton, click_count: int, source) -> None:
    """Click at a point. `click_count` >= 2 produces a proper multi-click by
    incrementing the click-state field on each successive down/up pair.
    """
    count = max(1, click_count)
    down_type, up_type = button.click_types

    smooth_move(point, source)
    _sleep_us(12_000)

    for n in range(1, count + 1):
        down = CGEventCreateMouseEvent(
            source, down_type, CGPointMake(point.x, point.y), button.cg_button
        )
        if down is not None:
            CGEventSetIntegerValueField(down, kCGMouseEventClickState, n)
        _post(down)
        _sleep_us(12_000)

        up = CGEventCreateMouseEvent(
            source, up_type, CGPointMake(point.x, point.y), button.cg_button
        )
        if up is not None:
            CGEventSetIntegerValueField(up, kCGMouseEventClickState, n)
        _post(up)
        _sleep_us(12_000)


def perform_mouse_down(point, button: MouseButton, source) -> None:
    smooth_move(point, source)
    _sleep_us(12_000)
    down = CGEventCreateMouseEvent(
        source, button.click_types[0], CGPointMake(point.x, point.y), button.cg_button
    )
    if down is not None:
        CGEventSetIntegerValueField(down, kCGMouseEventClickState, 1)
    _post(down)


def perform_mouse_up(point, button: MouseButton, source) -> None:
    up = CGEventCreateMouseEvent(
        source, button.click_types[1], CGPointMake(point.x, point.y), button.cg_button
    )
    if up is not None:
        CGEventSetIntegerValueField(up, kCGMouseEventClickState, 1)
    _post(up)


def perform_drag(from_point, to_point, button: MouseButton, source) -> None:
    """Press at `from`, drag through interpolated points, release at `to`."""
    down_type, up_type = button.click_types

    smooth_move(from_point, source)
    _sleep_us(20_000)

    down = CGEventCreateMouseEvent(
        source, down_type, CGPointMake(from_point.x, from_point.y), button.cg_button
    )
    if down is not None:
        CGEventSetIntegerValueField(down, kCGMouseEventClickState, 1)
    _post(down)
    _sleep_us(40_000)

    steps = 40
    for i in range(1, steps + 1):
        t = i / steps
        px = from_point.x + (to_point.x - from_point.x) * t
        py = from_point.y + (to_point.y - from_point.y) * t
        drag = CGEventCreateMouseEvent(source, button.drag_type, CGPointMake(px, py), button.cg_button)
        if drag is not None:
            CGEventSetIntegerValueField(drag, kCGMouseEventClickState, 1)
        _post(drag)
        _sleep_us(8_000)

    _sleep_us(30_000)
    up = CGEventCreateMouseEvent(
        source, up_type, CGPointMake(to_point.x, to_point.y), button.cg_button
    )
    if up is not None:
        CGEventSetIntegerValueField(up, kCGMouseEventClickState, 1)
    _post(up)


def natural_scrolling_enabled() -> bool:
    """Whether macOS "natural scrolling" is enabled (the default for trackpads)."""
    value = CFPreferencesCopyAppValue(
        "com.apple.swipescrolldirection", kCFPreferencesAnyApplication
    )
    if value is None:
        return False
    return bool(value)


def perform_scroll(
    direction: ScrollDirection,
    pages: int,
    lines_per_page: int,
    point,
    source,
) -> None:
    """Scroll `pages` pages in a direction, optionally after moving the cursor.

    `direction` is interpreted from the user's point of view regardless of the
    "natural scrolling" preference (we read it and flip the wheel signs).
    """
    if point is not None:
        smooth_move(point, source)
        _sleep_us(12_000)

    # Raw CoreGraphics convention (natural scrolling OFF): positive wheel1 reveals
    # content above ("up"), positive wheel2 reveals content to the left.
    flip = -1 if natural_scrolling_enabled() else 1
    total_lines = max(1, pages) * max(1, lines_per_page)
    for _ in range(total_lines):
        wheel1 = 0  # vertical
        wheel2 = 0  # horizontal
        if direction == ScrollDirection.UP:
            wheel1 = 1
        elif direction == ScrollDirection.DOWN:
            wheel1 = -1
        elif direction == ScrollDirection.LEFT:
            wheel2 = 1
        elif direction == ScrollDirection.RIGHT:
            wheel2 = -1
        # pyobjc's variadic wrapper takes exactly `wheelCount` wheel values.
        event = CGEventCreateScrollWheelEvent(
            source, kCGScrollEventUnitLine, 2, wheel1 * flip, wheel2 * flip
        )
        _post(event)
        _sleep_us(6_000)

File: 03_use_cases/13_macos_cua/src/metacua/Mouse.swift
import CoreGraphics
import Foundation

/// Which physical mouse button an action uses.
enum MouseButton: String {
  case left
  case right
  case center

  static func parse(_ raw: String) throws -> MouseButton {
    switch raw.lowercased() {
    case "left", "l", "primary": return .left
    case "right", "r", "secondary": return .right
    case "center", "middle", "c", "m": return .center
    default:
      throw CLIError("--mouse-button must be left|right|center (got '\(raw)')", code: 2)
    }
  }

  var cgButton: CGMouseButton {
    switch self {
    case .left: return .left
    case .right: return .right
    case .center: return .center
    }
  }

  /// Down / up event types for this button.
  var clickTypes: (down: CGEventType, up: CGEventType) {
    switch self {
    case .left: return (.leftMouseDown, .leftMouseUp)
    case .right: return (.rightMouseDown, .rightMouseUp)
    case .center: return (.otherMouseDown, .otherMouseUp)
    }
  }

  var dragType: CGEventType {
    switch self {
    case .left: return .leftMouseDragged
    case .right: return .rightMouseDragged
    case .center: return .otherMouseDragged
    }
  }
}

/// Vertical / horizontal scroll direction.
enum ScrollDirection: String {
  case up, down, left, right

  static func parse(_ raw: String) throws -> ScrollDirection {
    guard let d = ScrollDirection(rawValue: raw.lowercased()) else {
      throw CLIError("--direction must be up|down|left|right (got '\(raw)')", code: 2)
    }
    return d
  }
}

/// Shared event source for all generated events.
func eventSource() -> CGEventSource? {
  CGEventSource(stateID: .combinedSessionState)
}

private func post(_ event: CGEvent?) {
  event?.post(tap: .cghidEventTap)
}

/// Warp the cursor to a point in one event (instant; used per-step internally).
func moveMouse(to point: CGPoint, source: CGEventSource?) {
  post(
    CGEvent(
      mouseEventSource: source,
      mouseType: .mouseMoved,
      mouseCursorPosition: point,
      mouseButton: .left))
}

/// Current cursor location in global (top-left origin) coordinates.
func currentMouseLocation() -> CGPoint {
  CGEvent(source: nil)?.location ?? .zero
}

/// Glide the cursor to `target` along an eased path instead of teleporting, so
/// the motion looks smooth and human. Interpolates from the current position;
/// the step count scales with distance and is capped for responsiveness.
func smoothMove(to target: CGPoint, source: CGEventSource?) {
  let start = currentMouseLocation()
  let dx = target.x - start.x
  let dy = target.y - start.y
  let distance = (dx * dx + dy * dy).squareRoot()
  guard distance >= 1 else {
    moveMouse(to: target, source: source)
    return
  }
  let steps = max(12, min(60, Int(distance / 12)))
  for i in 1...steps {
    let t = Double(i) / Double(steps)
    let e = t < 0.5 ? 2 * t * t : 1 - pow(-2 * t + 2, 2) / 2 // ease-in-out
    moveMouse(to: CGPoint(x: start.x + dx * e, y: start.y + dy * e), source: source)
    usleep(6_000)
  }
}

/// Click at a point. `clickCount` >= 2 produces a proper multi-click by
/// incrementing the click-state field on each successive down/up pair, which is
/// what AppKit inspects to recognise double/triple clicks.
func performClick(
  at point: CGPoint,
  button: MouseButton,
  clickCount: Int,
  source: CGEventSource?
) {
  let count = max(1, clickCount)
  let types = button.clickTypes

  smoothMove(to: point, source: source)
  usleep(12_000)

  for n in 1...count {
    let down = CGEvent(
      mouseEventSource: source,
      mouseType: types.down,
      mouseCursorPosition: point,
      mouseButton: button.cgButton)
    down?.setIntegerValueField(.mouseEventClickState, value: Int64(n))
    post(down)
    usleep(12_000)

    let up = CGEvent(
      mouseEventSource: source,
      mouseType: types.up,
      mouseCursorPosition: point,
      mouseButton: button.cgButton)
    up?.setIntegerValueField(.mouseEventClickState, value: Int64(n))
    post(up)
    usleep(12_000)
  }
}

func performMouseDown(at point: CGPoint, button: MouseButton, source: CGEventSource?) {
  smoothMove(to: point, source: source)
  usleep(12_000)
  let down = CGEvent(
    mouseEventSource: source,
    mouseType: button.clickTypes.down,
    mouseCursorPosition: point,
    mouseButton: button.cgButton)
  down?.setIntegerValueField(.mouseEventClickState, value: 1)
  post(down)
}

func performMouseUp(at point: CGPoint, button: MouseButton, source: CGEventSource?) {
  let up = CGEvent(
    mouseEventSource: source,
    mouseType: button.clickTypes.up,
    mouseCursorPosition: point,
    mouseButton: button.cgButton)
  up?.setIntegerValueField(.mouseEventClickState, value: 1)
  post(up)
}

/// Press at `from`, drag through interpolated points, release at `to`.
func performDrag(
  from: CGPoint,
  to: CGPoint,
  button: MouseButton,
  source: CGEventSource?
) {
  let types = button.clickTypes

  smoothMove(to: from, source: source)
  usleep(20_000)

  let down = CGEvent(
    mouseEventSource: source,
    mouseType: types.down,
    mouseCursorPosition: from,
    mouseButton: button.cgButton)
  down?.setIntegerValueField(.mouseEventClickState, value: 1)
  post(down)
  usleep(40_000)

  let steps = 40
  for i in 1...steps {
    let t = Double(i) / Double(steps)
    let p = CGPoint(
      x: from.x + (to.x - from.x) * t,
      y: from.y + (to.y - from.y) * t)
    let drag = CGEvent(
      mouseEventSource: source,
      mouseType: button.dragType,
      mouseCursorPosition: p,
      mouseButton: button.cgButton)
    drag?.setIntegerValueField(.mouseEventClickState, value: 1)
    post(drag)
    usleep(8_000)
  }

  usleep(30_000)
  let up = CGEvent(
    mouseEventSource: source,
    mouseType: types.up,
    mouseCursorPosition: to,
    mouseButton: button.cgButton)
  up?.setIntegerValueField(.mouseEventClickState, value: 1)
  post(up)
}

/// Whether macOS "natural scrolling" is enabled (the default for trackpads).
/// When on, the system inverts the relationship between a wheel delta and the
/// direction content moves, so we flip our signs to keep `direction` meaning the
/// same thing (e.g. `down` always reveals content further down the document).
func naturalScrollingEnabled() -> Bool {
  if let value = CFPreferencesCopyAppValue(
    "com.apple.swipescrolldirection" as CFString,
    kCFPreferencesAnyApplication
  ) {
    return (value as? Bool) ?? false
  }
  return false
}

/// Scroll `pages` pages in a direction, optionally after moving the cursor to a
/// point first (so the scroll targets the element under that point).
///
/// `direction` is interpreted from the user's point of view: `down`/`right`
/// reveal content further down/right in the document, `up`/`left` reveal content
/// above/left — and this holds regardless of the "natural scrolling" preference
/// (we read it and flip the wheel signs to compensate).
///
/// Scrolling is emitted in small line-unit steps for smoothness and broad app
/// compatibility. `linesPerPage` controls how much one "page" moves.
func performScroll(
  direction: ScrollDirection,
  pages: Int,
  linesPerPage: Int,
  at point: CGPoint?,
  source: CGEventSource?
) {
  if let point = point {
    smoothMove(to: point, source: source)
    usleep(12_000)
  }

  // Raw CoreGraphics convention (natural scrolling OFF): positive wheel1
  // reveals content above ("up"), positive wheel2 reveals content to the left.
  // Flip when natural scrolling is ON so the user-facing direction is stable.
  let flip: Int32 = naturalScrollingEnabled() ? -1 : 1
  let totalLines = max(1, pages) * max(1, linesPerPage)
  for _ in 0..<totalLines {
    var wheel1: Int32 = 0 // vertical
    var wheel2: Int32 = 0 // horizontal
    switch direction {
    case .up: wheel1 = 1
    case .down: wheel1 = -1
    case .left: wheel2 = 1
    case .right: wheel2 = -1
    }
    let event = CGEvent(
      scrollWheelEvent2Source: source,
      units: .line,
      wheelCount: 2,
      wheel1: wheel1 * flip,
      wheel2: wheel2 * flip,
      wheel3: 0)
    post(event)
    usleep(6_000)
  }
}
